capsules/extra/src/eeprom.rs

```diff
@@ -0,0 +1,259 @@
+                self.buffer.replace(txbuffer);
+                self.i2c.disable();
+                Err((error.into(), buf))
+            } else {
+                self.client_page.replace(buf);
+                Ok(())
+            }
+        } else {
+            Err((ErrorCode::RESERVE, buf))
+        }
+    }
+
+    fn erase_sector(&self, page_number: usize) -> Result<(), ErrorCode> {
+        self.i2c.enable();
+        let address = page_number * PAGE_SIZE;
+        // Schedule page write and do first
+        if let Some(txbuffer) = self.buffer.take() {
+            txbuffer[0] = ((address >> 8) & 0x00ff) as u8;
+            txbuffer[1] = (address & 0x00ff) as u8;
+
+            let write_len = cmp::min(txbuffer.len() - 2, PAGE_SIZE);
+
+            for i in 0..write_len {
+                txbuffer[i + 2] = 0;
+            }
+
+            self.state.set(State::Erasing);
+            if let Err((error, txbuffer)) = self.i2c.write(txbuffer, write_len + 2) {
+                self.buffer.replace(txbuffer);
+                self.i2c.disable();
+                Err(error.into())
+            } else {
+                Ok(())
+            }
+        } else {
+            Err(ErrorCode::RESERVE)
+        }
+    }
+}
+
+impl I2CClient for EEPROM<'static> {
+    fn command_complete(&self, buffer: &'static mut [u8], status: Result<(), Error>) {
+        match self.state.get() {
+            State::Reading => {
+                self.state.set(State::Idle);
+                self.i2c.disable();
+                if let Some(client_page) = self.client_page.take() {
+                    for i in 0..PAGE_SIZE {
+                        client_page.0[i] = buffer[i];
+                    }
+                    self.buffer.replace(buffer);
+                    self.flash_client.map(|client| {
+                        if status.is_err() {
+                            client.read_complete(client_page, hil::flash::Error::FlashError);
+                        } else {
+                            client.read_complete(client_page, hil::flash::Error::CommandComplete);
+                        }
+                    });
+                } else {
+                    self.buffer.replace(buffer);
+                }
+            }
+            State::Writing => {
+                self.state.set(State::Idle);
+                self.buffer.replace(buffer);
+                self.i2c.disable();
+                self.flash_client.map(|client| {
+                    if let Some(client_page) = self.client_page.take() {
+                        if status.is_err() {
+                            client.write_complete(client_page, hil::flash::Error::FlashError);
+                        } else {
+                            client.write_complete(client_page, hil::flash::Error::CommandComplete);
+                        }
+                    } else {
+                        panic!("EEPROM Writing op completed without client_page");
```

Comment:
I don't think we need this panic. It's pretty clear that client_page will be there.

Reply:
same behaviour is in lines 194-196, but without panic, removed that